app/src/test/java/com/futurice/freesound/network/api/DefaultFreeSoundApiServiceTest.kt

```diff
@@ -0,0 +1,193 @@
+                TEST_CLIENT_ID,
+                TEST_CLIENT_SECRET)
+    }
+
+    @Test
+    fun getUser_emitsUser_whenApiSuccessful() {
+        arrange {
+            user { DUMMY_USER }
+        }
+
+        defaultFreeSoundApiService.getUser("username")
+                .test()
+                .assertValue(DUMMY_USER)
+    }
+
+    @Test
+    fun getUser_invokesApiWithUsernameParameter() {
+        arrange {
+            user { DUMMY_USER }
+        }
+        val username = "username"
+
+        defaultFreeSoundApiService.getUser(username).subscribe()
+
+
+        verify(freeSoundApi).user(username)
+    }
+
+    @Test
+    fun getUser_emitsError_whenApiError() {
+        arrange {
+            userError { ERROR }
+        }
+
+        defaultFreeSoundApiService.getUser("username")
+                .test()
+                .assertError(ERROR)
+    }
+
+    @Test
+    fun getAccessToken_emitsAccessToken_whenApiSuccessful() {
+        arrange {
+            token { DUMMY_ACCESS_TOKEN }
+        }
+        defaultFreeSoundApiService.getAccessToken("code")
+                .test()
+                .assertValue(DUMMY_ACCESS_TOKEN)
+    }
+
+    @Test
+    fun getAccessToken_invokesApiWithCorrectParameters() {
+        arrange {
+            token { DUMMY_ACCESS_TOKEN }
+        }
+        val code = "code"
+
+        defaultFreeSoundApiService.getAccessToken(code).subscribe()
+
+
+        verify(freeSoundApi).accessToken(TEST_CLIENT_ID,
+                TEST_CLIENT_SECRET,
+                ApiConstants.AUTHORIZATION_CODE_GRANT_TYPE_VALUE,
+                code)
+    }
+
+    @Test
+    fun getAccessToken_emitsError_whenApiError() {
+        arrange {
+            tokenError { ERROR }
+        }
+
+        defaultFreeSoundApiService.getAccessToken("code")
+                .test()
+                .assertError(ERROR)
+    }
+
+    @Test
+    fun search_emitsResults_whenApiSuccessful() {
+        arrange {
+            search { TEST_SEARCH_RESULT }
+        }
+
+        defaultFreeSoundApiService.search("query")
+                .test()
+                .assertValue(TEST_SEARCH_RESULT)
+    }
+
+    @Test
+    fun search_invokesApiWithCorrectParameters() {
+        arrange {
+            search { TEST_SEARCH_RESULT }
+        }
+        val query = "query"
+        defaultFreeSoundApiService.search(query).subscribe()
+
+        verify(freeSoundApi).search(eq(query), isNull<String>(), eq(SoundFields.BASE))
+    }
+
+    @Test
+    fun search_emitsError_whenApiError() {
+        arrange {
+            searchError { ERROR }
+        }
+
+        defaultFreeSoundApiService.search("query")
+                .test()
+                .assertError(ERROR)
+    }
+
+    fun arrange(init: Arrangement.() -> Unit) = Arrangement().apply(init)
+
+    inner class Arrangement {
+
+        fun search(init: () -> SoundSearchResult): Any =
+                `when`(freeSoundApi.search(anyString(), any<String>(), any(SoundFields::class.java)))
```

Comment:
Could we make this `any(SoundFields::class.java)` become:

```kotlin
    inline fun <reified T> any() : T = Mockito.any(T::class.java)
```

so call site is: ` any<SoundFields>()`

There's also this: https://github.com/nhaarman/mockito-kotlin/ which I has these already (I used this in kotlin-options).